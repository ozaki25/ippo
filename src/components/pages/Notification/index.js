import React from 'react';
import { Divider, Typography } from '@material-ui/core';
import propTypes from 'prop-types';
import Container from 'src/components/templates/Container';
import dateFormat from 'src/utils/dateFormat';

class Notification extends React.Component {
  async componentDidMount() {
    const {
      user: {
        refetch,
        variables: { uid },
      },
      match: {
        params: { id },
      },
      readNotification,
    } = this.props;
    const {
      data: {
        fetchUser: { notifications },
      },
    } = await refetch({ uid });
    const notification = notifications.find(n => n.id === id);
    if (notification.checked) return;
    readNotification({ variables: { uid, notificationId: id } });
  }

  render() {
    const {
      user: { fetchUser },
      authUser,
      history,
      match: {
        params: { id },
      },
      firebase,
    } = this.props;
    const notification = fetchUser ? fetchUser.notifications.find(n => n.id === id) : null;
    return (
      <Container title="通知" back authUser={authUser} history={history} firebase={firebase}>
        {notification && (
          <>
            <Typography variant="h6">{notification.title}</Typography>
            <Typography color="textSecondary">
              {dateFormat.datetimeJa(notification.timestamp)}
            </Typography>
            <Typography paragraph>{notification.content}</Typography>
            <Divider />
          </>
        )}
      </Container>
    );
  }
}

Notification.displayName = 'Notification';

Notification.propTypes = {
  user: propTypes.shape({
    fetchUser: propTypes.shape({
      notifications: propTypes.arrayOf(
        propTypes.shape({
          id: propTypes.string,
          checked: propTypes.bool,
          title: propTypes.string,
          content: propTypes.string,
          timestamp: propTypes.string,
        }),
      ),
    }),
    refetch: propTypes.func.isRequired,
    variables: propTypes.shape({
      uid: propTypes.string.isRequired,
    }).isRequired,
  }),
  match: propTypes.shape({
    params: propTypes.shape({
      id: propTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  readNotification: propTypes.func.isRequired,
  authUser: propTypes.shape({
    displayName: propTypes.string.isRequired,
    uid: propTypes.string.isRequired,
  }).isRequired,
  history: propTypes.shape({
    push: propTypes.func.isRequired,
    goBack: propTypes.func.isRequired,
    replace: propTypes.func.isRequired,
  }).isRequired,
};

Notification.defaultProps = {};

export default Notification;
