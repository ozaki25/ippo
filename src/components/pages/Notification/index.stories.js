import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import Notification from '.';

const stories = storiesOf('pages/Notification', module);

const props = {
  publishNotification: action('publish'),
  registerNotification: action('register'),
};

stories.add('通常パターン', () => <Notification {...props} />);
