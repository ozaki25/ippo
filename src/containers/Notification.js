import React from 'react';
import { useQuery, useMutation } from '@apollo/client';
import { compose } from 'recompose';
import { useHistory } from 'react-router-dom';

import useFirebase from 'src/hooks/useFirebase';
import { withAuthorization } from 'src/hoc/Sessions';
import query from 'src/graphql/query';
import mutation from 'src/graphql/mutation';
import Notification from 'src/components/pages/Notification';

const WithNotification = compose(withAuthorization)(NotificationContainer);

function NotificationContainer(props) {
  const { uid } = props.authUser;
  const history = useHistory();
  const firebase = useFirebase();
  const { data, loading, error, refetch, variables } = useQuery(
    query.fetchUser,
    {
      variables: { uid },
    },
  );
  const [readNotification] = useMutation(mutation.readNotification);
  return (
    <Notification
      {...props}
      history={history}
      firebase={firebase}
      readNotification={readNotification}
      user={{ ...data, loading, error, refetch, variables }}
    />
  );
}

export default WithNotification;
