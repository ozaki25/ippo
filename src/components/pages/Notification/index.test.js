import React from 'react';
import { snapshot } from 'test/helpers';
import Notification from '.';

const props = {
  publishNotification: jest.fn(),
  registerNotification: jest.fn(),
};

snapshot('Notification/nomal', <Notification {...props} />);
