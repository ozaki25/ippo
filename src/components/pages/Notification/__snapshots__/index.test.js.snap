// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot: Notification/nomal 1`] = `
.c0 {
  padding: 10px 15px;
}

.c1 {
  margin: 15px 0;
}

<MockRouter
  createHref={[Function]}
  location=""
  params={Object {}}
  path=""
  push={[Function]}
  replace={[Function]}
  url=""
>
  <Notification
    publishNotification={[MockFunction]}
    registerNotification={[MockFunction]}
  >
    <Container
      authUser={null}
      noPadding={false}
    >
      <withRouter()
        authUser={null}
      >
        <Route>
          <Component
            authUser={null}
            history={
              Object {
                "createHref": [Function],
                "path": "",
                "push": [Function],
                "replace": [Function],
              }
            }
            location=""
            match={
              Object {
                "params": Object {},
                "path": "",
                "url": "",
              }
            }
          >
            <NavigationBar
              authUser={null}
              firebase={null}
              history={
                Object {
                  "createHref": [Function],
                  "path": "",
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location=""
              match={
                Object {
                  "params": Object {},
                  "path": "",
                  "url": "",
                }
              }
            >
              <Blueprint3.Navbar>
                <div
                  className="bp3-navbar"
                >
                  <Blueprint3.NavbarGroup
                    align="left"
                  >
                    <div
                      className="bp3-navbar-group bp3-align-left"
                    >
                      <Blueprint3.NavbarHeading>
                        <div
                          className="bp3-navbar-heading"
                        >
                          IPPO
                        </div>
                      </Blueprint3.NavbarHeading>
                      <Blueprint3.NavbarDivider>
                        <div
                          className="bp3-navbar-divider"
                        />
                      </Blueprint3.NavbarDivider>
                    </div>
                  </Blueprint3.NavbarGroup>
                  <Blueprint3.NavbarGroup
                    align="right"
                  >
                    <div
                      className="bp3-navbar-group bp3-align-right"
                    >
                      <Blueprint3.Button
                        icon="notifications"
                        minimal={true}
                        onClick={[Function]}
                      >
                        <button
                          className="bp3-button bp3-minimal"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          onKeyUp={[Function]}
                          type="button"
                        >
                          <Blueprint3.Icon
                            icon="notifications"
                            key="leftIcon"
                          >
                            <span
                              className="bp3-icon bp3-icon-notifications"
                              icon="notifications"
                            >
                              <svg
                                data-icon="notifications"
                                height={16}
                                viewBox="0 0 16 16"
                                width={16}
                              >
                                <desc>
                                  notifications
                                </desc>
                                <path
                                  d="M8 16c1.1 0 2-.9 2-2H6c0 1.1.9 2 2 2zm6-5c-.55 0-1-.45-1-1V6c0-2.43-1.73-4.45-4.02-4.9 0-.04.02-.06.02-.1 0-.55-.45-1-1-1S7 .45 7 1c0 .04.02.06.02.1A4.992 4.992 0 0 0 3 6v4c0 .55-.45 1-1 1s-1 .45-1 1 .45 1 1 1h12c.55 0 1-.45 1-1s-.45-1-1-1z"
                                  fillRule="evenodd"
                                  key="0"
                                />
                              </svg>
                            </span>
                          </Blueprint3.Icon>
                          <Blueprint3.Icon
                            key="rightIcon"
                          />
                        </button>
                      </Blueprint3.Button>
                    </div>
                  </Blueprint3.NavbarGroup>
                </div>
              </Blueprint3.Navbar>
            </NavigationBar>
          </Component>
        </Route>
      </withRouter()>
      <styled.div
        noPadding={false}
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
  padding: ",
                  [Function],
                  ";
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          noPadding={false}
        >
          <div
            className="c0"
          >
            <Component>
              <h3
                className="bp3-heading"
              >
                プッシュ通知の管理
              </h3>
            </Component>
            <Blueprint3.Text>
              <div
                className=""
              >
                ※ボタンを押すと購読者全員に通知が送信されます。
              </div>
            </Blueprint3.Text>
            <Blueprint3.Button
              onClick={[Function]}
              text="プッシュ通知を送信する"
            >
              <button
                className="bp3-button"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <Blueprint3.Icon
                  key="leftIcon"
                />
                <span
                  className="bp3-button-text"
                  key="text"
                >
                  プッシュ通知を送信する
                </span>
                <Blueprint3.Icon
                  key="rightIcon"
                />
              </button>
            </Blueprint3.Button>
            <Blueprint3.Button
              onClick={[Function]}
              text="プッシュ通知を購読する"
            >
              <button
                className="bp3-button"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <Blueprint3.Icon
                  key="leftIcon"
                />
                <span
                  className="bp3-button-text"
                  key="text"
                >
                  プッシュ通知を購読する
                </span>
                <Blueprint3.Icon
                  key="rightIcon"
                />
              </button>
            </Blueprint3.Button>
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  margin: 15px 0;
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="c1"
                >
                  <Link
                    replace={false}
                    to="/"
                  >
                    <a
                      onClick={[Function]}
                    >
                      イベント一覧へ
                    </a>
                  </Link>
                </div>
              </StyledComponent>
            </styled.div>
          </div>
        </StyledComponent>
      </styled.div>
    </Container>
  </Notification>
</MockRouter>
`;

exports[`should match snapshot: Publish/nomal 1`] = `
.c0 {
  padding: 10px 15px;
}

.c1 {
  margin: 15px 0;
}

<MockRouter
  createHref={[Function]}
  location=""
  params={Object {}}
  path=""
  push={[Function]}
  replace={[Function]}
  url=""
>
  <Publish
    publishNotification={[MockFunction]}
  >
    <Container
      authUser={null}
      noPadding={false}
    >
      <withRouter()
        authUser={null}
      >
        <Route>
          <Component
            authUser={null}
            history={
              Object {
                "createHref": [Function],
                "path": "",
                "push": [Function],
                "replace": [Function],
              }
            }
            location=""
            match={
              Object {
                "params": Object {},
                "path": "",
                "url": "",
              }
            }
          >
            <NavigationBar
              authUser={null}
              firebase={null}
              history={
                Object {
                  "createHref": [Function],
                  "path": "",
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location=""
              match={
                Object {
                  "params": Object {},
                  "path": "",
                  "url": "",
                }
              }
            >
              <Blueprint3.Navbar>
                <div
                  className="bp3-navbar"
                >
                  <Blueprint3.NavbarGroup
                    align="left"
                  >
                    <div
                      className="bp3-navbar-group bp3-align-left"
                    >
                      <Blueprint3.NavbarHeading>
                        <div
                          className="bp3-navbar-heading"
                        >
                          IPPO
                        </div>
                      </Blueprint3.NavbarHeading>
                      <Blueprint3.NavbarDivider>
                        <div
                          className="bp3-navbar-divider"
                        />
                      </Blueprint3.NavbarDivider>
                    </div>
                  </Blueprint3.NavbarGroup>
                  <Blueprint3.NavbarGroup
                    align="right"
                  >
                    <div
                      className="bp3-navbar-group bp3-align-right"
                    >
                      <Blueprint3.Button
                        icon="wrench"
                        minimal={true}
                        onClick={[Function]}
                      >
                        <button
                          className="bp3-button bp3-minimal"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          onKeyUp={[Function]}
                          type="button"
                        >
                          <Blueprint3.Icon
                            icon="wrench"
                            key="leftIcon"
                          >
                            <span
                              className="bp3-icon bp3-icon-wrench"
                              icon="wrench"
                            >
                              <svg
                                data-icon="wrench"
                                height={16}
                                viewBox="0 0 16 16"
                                width={16}
                              >
                                <desc>
                                  wrench
                                </desc>
                                <path
                                  d="M15.83 3.7l-3.06 3.05-2.84-.7-.7-2.83L12.29.17a5.004 5.004 0 0 0-4.83 1.29 4.967 4.967 0 0 0-1.12 5.36L.58 12.58c-.36.36-.58.86-.58 1.41 0 1.1.9 2 2 2 .55 0 1.05-.22 1.41-.59l5.77-5.77c1.79.69 3.91.33 5.35-1.12 1.32-1.3 1.74-3.15 1.3-4.81z"
                                  fillRule="evenodd"
                                  key="0"
                                />
                              </svg>
                            </span>
                          </Blueprint3.Icon>
                          <Blueprint3.Icon
                            key="rightIcon"
                          />
                        </button>
                      </Blueprint3.Button>
                    </div>
                  </Blueprint3.NavbarGroup>
                </div>
              </Blueprint3.Navbar>
            </NavigationBar>
          </Component>
        </Route>
      </withRouter()>
      <styled.div
        noPadding={false}
      >
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-bdVaJa",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "
  padding: ",
                  [Function],
                  ";
",
                ],
              },
              "displayName": "styled.div",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-bdVaJa",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
          noPadding={false}
        >
          <div
            className="c0"
          >
            <Component>
              <h3
                className="bp3-heading"
              >
                プッシュ通知の管理
              </h3>
            </Component>
            <Blueprint3.Text>
              <div
                className=""
              >
                ※ボタンを押すと購読者全員に通知が送信されます。
              </div>
            </Blueprint3.Text>
            <Blueprint3.Button
              onClick={[Function]}
              text="プッシュ通知を送信する"
            >
              <button
                className="bp3-button"
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <Blueprint3.Icon
                  key="leftIcon"
                />
                <span
                  className="bp3-button-text"
                  key="text"
                >
                  プッシュ通知を送信する
                </span>
                <Blueprint3.Icon
                  key="rightIcon"
                />
              </button>
            </Blueprint3.Button>
            <styled.div>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bwzfXH",
                      "isStatic": true,
                      "lastClassName": "c1",
                      "rules": Array [
                        "
  margin: 15px 0;
",
                      ],
                    },
                    "displayName": "styled.div",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "sc-bwzfXH",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="c1"
                >
                  <Link
                    replace={false}
                    to="/events"
                  >
                    <a
                      onClick={[Function]}
                    >
                      イベント一覧へ
                    </a>
                  </Link>
                </div>
              </StyledComponent>
            </styled.div>
          </div>
        </StyledComponent>
      </styled.div>
    </Container>
  </Publish>
</MockRouter>
`;
